pkg/operator/csi/csidrivercontroller/csi_driver_controller.go

```diff
@@ -0,0 +1,369 @@
+func NewCSIDriverController(
+	name string,
+	csiDriverName string,
+	csiDriverNamespace string,
+	operatorClient v1helpers.OperatorClient,
+	assetFunc func(string) []byte,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+) *CSIDriverController {
+	c := &CSIDriverController{
+		name:               name,
+		csiDriverName:      csiDriverName,
+		csiDriverNamespace: csiDriverNamespace,
+		kubeClient:         kubeClient,
+		operatorClient:     operatorClient,
+		assetFunc:          assetFunc,
+		images:             imagesFromEnv(),
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), csiDriverName),
+		eventRecorder:      eventRecorder.WithComponentSuffix("csi-driver-controller-recorder-" + strings.ToLower(name)),
+	}
+	operatorClient.Informer().AddEventHandler(c.eventHandler(csiDriverName))
+	c.informersSynced = append(c.informersSynced, operatorClient.Informer().HasSynced)
+	return c
+}
+
+// WithControllerService returns a controller with the CSI Controller Service.
+// The CSI Controller Service is represented by a Deployment.
+func (c *CSIDriverController) WithControllerService(informer appsinformersv1.DeploymentInformer, file string) *CSIDriverController {
+	informer.Informer().AddEventHandler(c.eventHandler("deployment"))
+	c.informersSynced = append(c.informersSynced, informer.Informer().HasSynced)
+	c.controllerManifest = c.assetFunc(file)
+	return c
+}
+
+// WithNodeService returns a controller with the CSI Node Service.
+// The CSI Node Service is represented by a DaemonSet.
+func (c *CSIDriverController) WithNodeService(informer appsinformersv1.DaemonSetInformer, file string) *CSIDriverController {
+	informer.Informer().AddEventHandler(c.eventHandler("daemonSet"))
+	c.informersSynced = append(c.informersSynced, informer.Informer().HasSynced)
+	c.nodeManifest = c.assetFunc(file)
+	return c
+}
+
+// Run starts syncing the controller.
+func (c *CSIDriverController) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	if c.nodeManifest == nil {
+		klog.Errorf("Controller doesn't have a CSI Node Service, quitting")
+		return
+	}
+
+	stopCh := ctx.Done()
+	if !cache.WaitForCacheSync(stopCh, c.informersSynced...) {
+		return
+	}
+	klog.Infof("Caches synced, running the controller")
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *CSIDriverController) sync() error {
+	meta, err := c.operatorClient.GetObjectMeta()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.Warningf("Object metadata not found: %v", err)
+			return nil
+		}
+		return err
+	}
+
+	opSpec, opStatus, opResourceVersion, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	// We only support Managed for now
+	if opSpec.ManagementState != opv1.Managed {
+		return nil
+	}
+
+	startTime := time.Now()
+	klog.Info("Starting syncing operator at ", startTime)
```

Comment:
I recommend V(4).  You'll see it when you crank up logs, but you won't get noise during normal running.

Reply:
bumped to 4